ec/src/pairing.rs

```diff
@@ -0,0 +1,331 @@
+    fn sub_assign(&mut self, other: &'a Self) {
+        self.0 *= other.0.cyclotomic_inverse().unwrap();
+    }
+}
+
+impl<'a, P: Pairing> Sub<&'a Self> for PairingOutput<P> {
+    type Output = Self;
+
+    #[inline]
+    fn sub(mut self, other: &'a Self) -> Self {
+        self -= other;
+        self
+    }
+}
+
+ark_ff::impl_additive_ops_from_ref!(PairingOutput, Pairing);
+
+impl<P: Pairing, T: Borrow<P::ScalarField>> MulAssign<T> for PairingOutput<P> {
+    fn mul_assign(&mut self, other: T) {
+        *self = self.mul_bigint(other.borrow().into_bigint());
+    }
+}
+
+impl<P: Pairing, T: Borrow<P::ScalarField>> Mul<T> for PairingOutput<P> {
+    type Output = Self;
+
+    fn mul(self, other: T) -> Self {
+        self.mul_bigint(other.borrow().into_bigint())
+    }
+}
+
+impl<P: Pairing> Zeroize for PairingOutput<P> {
+    fn zeroize(&mut self) {
+        self.0.zeroize()
+    }
+}
+
+impl<P: Pairing> Neg for PairingOutput<P> {
+    type Output = Self;
+
+    #[inline]
+    fn neg(self) -> Self {
+        Self(self.0.cyclotomic_inverse().unwrap())
+    }
+}
+
+impl<P: Pairing> Distribution<PairingOutput<P>> for Standard {
+    #[inline]
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> PairingOutput<P> {
+        // Sample a random G1 element
+        let g1 = P::G1::rand(rng);
+        // Sample a random G2 element
+        let g2 = P::G2::rand(rng);
+        P::pairing(g1, g2)
+    }
+}
+
+impl<P: Pairing> Group for PairingOutput<P> {
+    type ScalarField = P::ScalarField;
+
+    fn generator() -> Self {
+        // TODO: hardcode these values.
+        // Sample a random G1 element
+        let g1 = P::G1::generator();
+        // Sample a random G2 element
+        let g2 = P::G2::generator();
+        P::pairing(g1.into(), g2.into())
+    }
+
+    fn double_in_place(&mut self) -> &mut Self {
+        self.0.cyclotomic_square_in_place();
+        self
+    }
+
+    fn mul_bigint(&self, other: impl AsRef<[u64]>) -> Self {
+        Self(self.0.cyclotomic_exp(other.as_ref()))
+    }
+
+    fn mul_bits_be(&self, other: impl Iterator<Item = bool>) -> Self {
+        // Convert back from bits to [u64] limbs
+        let other = other
+            .collect::<Vec<_>>()
+            .chunks(64)
+            .map(|chunk| {
+                chunk
+                    .iter()
+                    .enumerate()
+                    .fold(0, |r, (i, bit)| r | u64::from(*bit) << i)
+            })
+            .collect::<Vec<_>>();
+        Self(self.0.cyclotomic_exp(&other))
+    }
+}
+
+/// Represents the output of the Miller loop of the pairing.
+#[derive(Derivative)]
+#[derivative(
+    Copy(bound = "P: Pairing"),
+    Clone(bound = "P: Pairing"),
+    Debug(bound = "P: Pairing"),
+    PartialEq(bound = "P: Pairing"),
+    Eq(bound = "P: Pairing"),
+    PartialOrd(bound = "P: Pairing"),
+    Ord(bound = "P: Pairing")
+)]
+#[must_use]
+pub struct MillerLoopOutput<P: Pairing>(pub P::TargetField);
```

Comment:
Are you wrapping `TargetField` in another struct just to be able to avoid Rust's orphan rule, or is there an underlying API design decision?
It seems somewhat awkward that we're mixing a very high-level API names  (`MillerLoopOutput`) with names of algebraic structures in the miller loop method signature. For consistency, I'd suggest either:
```
    fn multi_miller_loop(
        a: impl IntoIterator<Item = impl Into<Self::MillerLoopInputA>>,
        b: impl IntoIterator<Item = impl Into<Self::MillerLoopInputB>>,
    ) -> MillerLoopOutput<Self>;
```
or something along the lines of:
```
    fn multi_miller_loop(
        a: ...G1Prepared...,
        b: ...G2Prepared...,
    ) -> FqkEqClass;
```
of which I'd go for the latter, though maybe `FqkEqClass` is a bit too explicit? I don't think any of our cuves implement the Weil pairing though, so that should always be the output of the miller loop.

Same for `PairingOutput`, maybe `Gt` instead?

Reply:
I think `FqkEqClass` is unlikely going to be used in another place, and for one-time use, I prefer to dismiss it.

The output is an element in `Fqk`, we just need to do a final exp after that. Note that we do not always use it just for equality check. Pairing could also be used, without a equality check in the middle, for **group commitments**.

For this reason, I would not recommend `FqkEqClass`.

Reply:
By the way, I think dedicating a struct for this is good `MillerLoopOutput` as it prevents developers from forgetting to do final exp.

For this purpose, I recommend we drop the derivation of `PartialEq, Eq, PartialOrd, Ord`.

Reply:
I'd be fine with `MillerLoopInput1`, but that it seems a bit less informative than `G1Prepared`.

The reason to have `MillerLoopOutput` be not just `Self::TargetField` is to make it explicit that it's not meant to be used by itself.